File: python/src/features.py
"""
Features partagees entre l'ENTRAINEMENT et l'INFERENCE.

Source de verite unique : train_pipeline.py et main.py importent ces fonctions.
Si ces implementations divergent entre train et inference, la reconnaissance
s'effondre (vecteurs incomparables). Ne pas dupliquer ailleurs.

Portable C++ ligne a ligne (numpy + cv2 -> Eigen/OpenCV C++).
"""

import cv2
import numpy as np


def crop_to_content(binary):
    """Recadre sur la bounding box du trait."""
    ys, xs = np.where(binary > 0)
    if len(xs) == 0:
        return binary
    return binary[ys.min():ys.max() + 1, xs.min():xs.max() + 1]


def normalize_crop(binary, target_size=32):
    """Centre le glyphe dans un carre (preserve ratio) puis redimensionne."""
    content = crop_to_content(binary)
    h, w = content.shape
    if h == 0 or w == 0:
        return np.zeros((target_size, target_size), dtype=np.uint8)
    side = max(h, w)
    square = np.zeros((side, side), dtype=np.uint8)
    yo, xo = (side - h) // 2, (side - w) // 2
    square[yo:yo + h, xo:xo + w] = content
    return cv2.resize(square, (target_size, target_size),
                      interpolation=cv2.INTER_NEAREST)


def compute_hog(img, cell_size=8, n_bins=9):
    """
    HOG maison. img de taille fixe (ex 32x32).
    Gradients Sobel sur l'image entiere, orientation non signee [0, pi),
    histogramme pondere par la magnitude par cellule, normalisation L2.
    """
    gx = cv2.Sobel(img.astype(np.float32), cv2.CV_32F, 1, 0, ksize=3)
    gy = cv2.Sobel(img.astype(np.float32), cv2.CV_32F, 0, 1, ksize=3)
    magnitude = np.sqrt(gx ** 2 + gy ** 2)
    orientation = np.arctan2(gy, gx) % np.pi

    h, w = img.shape
    bin_width = np.pi / n_bins
    features = []
    for i in range(0, h - cell_size + 1, cell_size):
        for j in range(0, w - cell_size + 1, cell_size):
            cell_mag = magnitude[i:i + cell_size, j:j + cell_size].ravel()
            cell_ori = orientation[i:i + cell_size, j:j + cell_size].ravel()
            hist = np.zeros(n_bins, dtype=np.float32)
            for m, o in zip(cell_mag, cell_ori):
                b = int(o / bin_width)
                if b >= n_bins:
                    b = n_bins - 1
                hist[b] += m
            features.append(hist)

    features = np.concatenate(features)
    norm = np.linalg.norm(features)
    if norm > 0:
        features = features / norm
    return features.astype(np.float32)

File: cpp/src/features.cc
#include "pyplus/features.hh"

#include <cmath>
#include <stdexcept>

#include <opencv2/imgproc.hpp>

namespace pyplus
{

cv::Mat crop_to_content(const cv::Mat& binary)
{
    std::vector<cv::Point> nonzero;
    cv::findNonZero(binary, nonzero);
    if (nonzero.empty())
        return binary;
    const cv::Rect bbox = cv::boundingRect(nonzero);
    return binary(bbox);
}

cv::Mat normalize_crop(const cv::Mat& binary, int target_size)
{
    const cv::Mat content = crop_to_content(binary);
    const int h = content.rows;
    const int w = content.cols;
    if (h == 0 || w == 0)
        return cv::Mat::zeros(target_size, target_size, CV_8U);

    const int side = std::max(h, w);
    cv::Mat square = cv::Mat::zeros(side, side, CV_8U);
    const int yo = (side - h) / 2;
    const int xo = (side - w) / 2;
    content.copyTo(square(cv::Rect(xo, yo, w, h)));

    cv::Mat resized;
    cv::resize(square, resized, cv::Size(target_size, target_size), 0, 0,
               cv::INTER_NEAREST);
    return resized;
}

std::vector<float> compute_hog(const cv::Mat& img, int cell_size, int n_bins)
{
    if (img.empty() || img.type() != CV_8U)
        throw std::runtime_error("compute_hog attend une image uint8");

    cv::Mat img_f;
    img.convertTo(img_f, CV_32F);
    cv::Mat gx, gy;
    cv::Sobel(img_f, gx, CV_32F, 1, 0, 3);
    cv::Sobel(img_f, gy, CV_32F, 0, 1, 3);

    const int h = img.rows;
    const int w = img.cols;
    const float pi = static_cast<float>(CV_PI);
    const float bin_width = pi / static_cast<float>(n_bins);

    std::vector<float> features;
    for (int i = 0; i + cell_size <= h; i += cell_size)
    {
        for (int j = 0; j + cell_size <= w; j += cell_size)
        {
            std::vector<float> hist(static_cast<std::size_t>(n_bins), 0.0f);
            for (int y = i; y < i + cell_size; ++y)
            {
                const float* gx_row = gx.ptr<float>(y);
                const float* gy_row = gy.ptr<float>(y);
                for (int x = j; x < j + cell_size; ++x)
                {
                    const float dx = gx_row[x];
                    const float dy = gy_row[x];
                    const float magnitude = std::sqrt(dx * dx + dy * dy);
                    // equivalent de numpy `arctan2(gy, gx) % pi` : [0, pi)
                    // gradient horizontal -> angle 0 exact : atan2(0, x<0)
                    // vaut pi-1ulp sur macOS mais pi pile sur Linux, ce qui
                    // fait basculer le bin (8 vs 0)
                    float orientation = 0.0f;
                    if (dy != 0.0f)
                    {
                        orientation = std::fmod(std::atan2(dy, dx), pi);
                        if (orientation < 0.0f)
                            orientation += pi;
                    }
                    int b = static_cast<int>(orientation / bin_width);
                    if (b >= n_bins)
                        b = n_bins - 1;
                    hist[static_cast<std::size_t>(b)] += magnitude;
                }
            }
            features.insert(features.end(), hist.begin(), hist.end());
        }
    }

    double norm_sq = 0.0;
    for (float v : features)
        norm_sq += static_cast<double>(v) * v;
    const float norm = static_cast<float>(std::sqrt(norm_sq));
    if (norm > 0.0f)
        for (float& v : features)
            v /= norm;
    return features;
}

} // namespace pyplus
